### DISCUSSION TITLE
How to access struct field of kprobe kernel function argument

### DISCUSSION THREAD

**Discussion by vincentmli** [bpftrace/bpftrace#2761] [Q&A] [answered]
Hi

 the bpftrace script I have below in the comment and the script explain what I want,  the first argument  is a `struct bpf_tramp_image *`, and  field member `struct bpf_ksym ksym` has field `name` which is what I want to print.

```
#!/usr/bin/env bpftrace

/*
 * include/linux/bpf.h
 * struct bpf_ksym {
 *       unsigned long            start;
 *       unsigned long            end;
 *       char                     name[KSYM_NAME_LEN]; #<---print out the name
 *  ...SNIP...
 *  };
 *
 * struct bpf_tramp_image {
 *       void *image;
 *       struct bpf_ksym ksym;
 *       struct percpu_ref pcref;
 *       void *ip_after_call;
 *       void *ip_epilogue;
 *  ....SNIP...
 *  };
 *
 * arch/x86/net/bpf_jit_com.c
 * int arch_prepare_bpf_trampoline(struct bpf_tramp_image *im, void *image, void *image_end,
 *                               const struct btf_func_model *m, u32 flags,
 *                               struct bpf_tramp_links *tlinks,
 *                               void *func_addr);
 *
 *
 */

BEGIN
{
        printf("Tracing trampoline image kernel symbol name. Hit Ctrl-C to end.\n");
}

kprobe:arch_prepare_bpf_trampoline
{

$im = (struct bpf_tramp_image *)arg0;

$name = $im.ksym->name;

printf("ksym name: %s\n", $name);

}
```

when I run above bpftrace script, I got error

```
./trampoline.bt:40:9-13: ERROR: Can not access field 'ksym' on type 'struct bpf_tramp_image *'. Try dereferencing it first, or using '->'
$name = $im.ksym->name;
        ~~~~
```
I think I am confused on how to access struct field pointer or non-pointer members, any help is greatly appreciated !

**Comment by vincentmli** [comment]
if I try following

```
kprobe:arch_prepare_bpf_trampoline
{

//$im = (struct bpf_tramp_image *)arg0;

//$name = ((*$im).ksym->name;

printf("ksym name: %s\n", ((struct bpf_tramp_image *)arg0)->ksym.name);

}
```
I got error:


`error: <unknown>:0:0: in function kprobe:arch_prepare_bpf_trampoline i64 (i8*): Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.`

**Reply by viktormalik** [reply, marked as answer]
This is the correct way to access the name. The problem is that `KSYM_NAME_LEN` is 512 and bpftrace tries to copy the entire string onto the BPF stack (which is also 512B).

The only way to workaround that at this moment is to redefine the type layout:
```
struct my_bpf_ksym {
    unsigned long start;
    unsigned long end;
    char name[64];
}

struct my_bpf_tramp_image {
    void *image;
    struct my_bpf_ksym ksym;
}

kprobe:arch_prepare_bpf_trampoline
{
    printf("ksym name: %s\n", ((struct my_bpf_tramp_image *)arg0)->ksym.name);
}
```
or in a shorter version:
```
struct my_bpf_tramp_image {
    void *image;
    unsigned long start;
    unsigned long end;
    char name[64];
}

kprobe:arch_prepare_bpf_trampoline
{
    printf("ksym name: %s\n", ((struct my_bpf_tramp_image *)arg0)->name);
}
```

This problem should be resolved by storing printf arguments (or in general strings) off-stack, we have some unfinished PRs for that (#1360, #1801) and I think that @xh4n3 was planning to pick up some of that work.

An alternative would be to tell bpftrace to just read a part of that string, but that's not possible at this point, either. `str` has an optional argument to specify the number of characters to read, but `str` accepts `char *` arguments instead of `char []`, so we'd need to implement the `&` (*address-of*) operator to convert `name` to `char *` (#553).

**Comment by vincentmli** [comment]
thank you for the detail explanation, really helpful :)
